mesonbuild/build.py

```diff
@@ -1067,15 +1067,23 @@ def extract_all_objects(self, recursive: bool = True) -> ExtractedObjects:
         return ExtractedObjects(self, self.sources, self.generated, self.objects,
                                 recursive, pch=True)
 
+    def get_toplevel_link_deps(self) -> ImmutableListProtocol[BuildTargetTypes]:
+        return self.link_targets
+
     def get_all_link_deps(self) -> ImmutableListProtocol[BuildTargetTypes]:
         return self.get_transitive_link_deps()
 
     @lru_cache(maxsize=None)
     def get_transitive_link_deps(self) -> ImmutableListProtocol[BuildTargetTypes]:
-        result: T.List[Target] = []
-        for i in self.link_targets:
-            result += i.get_all_link_deps()
-        return result
+        result: T.Set[Target] = []
+        stack: T.Deque[Target] = deque()
+        stack.appendleft(self)
+        while stack:
+            for i in stack.pop().get_toplevel_link_deps():
```

Comment:
i can be a static library that se need to skip here. I think you could call `i.links_dynamically()` because that would include shared libraries built with a custom_target().

Reply:
Wouldn't that change behavior from what master is doing? My goal here was only to make it faster, not to improve correctness.